stripe/src/main/java/com/stripe/android/AbstractEphemeralKey.java

```diff
@@ -200,66 +211,46 @@ String getType() {
         return mType;
     }
 
-    public static final Parcelable.Creator<EphemeralKey> CREATOR
-            = new Parcelable.Creator<EphemeralKey>() {
-
-        @Override
-        public EphemeralKey createFromParcel(Parcel in) {
-            return new EphemeralKey(in);
-        }
-
-        @Override
-        public EphemeralKey[] newArray(int size) {
-            return new EphemeralKey[size];
-        }
-    };
-
     @Nullable
-    static EphemeralKey fromString(@Nullable String rawJson) {
+    protected static <TEphemeralKey extends AbstractEphemeralKey> TEphemeralKey
+    fromString(@Nullable String rawJson, Class ephemeralKeyClass) {
         if (rawJson == null) {
             return null;
         }
 
         try {
             JSONObject object = new JSONObject(rawJson);
-            return fromJson(object);
+            return fromJson(object, ephemeralKeyClass);
         } catch (JSONException ignored) {
             return null;
         }
     }
 
     @Nullable
-    static EphemeralKey fromJson(@Nullable JSONObject jsonObject) {
+    @SuppressWarnings({"checkstyle:LineLength", "checkstyle:IllegalCatch"})
+    protected static <TEphemeralKey extends AbstractEphemeralKey> TEphemeralKey
+    fromJson(@Nullable JSONObject jsonObject, Class ephemeralKeyClass) {
         if (jsonObject == null) {
             return null;
         }
 
         try {
-            long created = jsonObject.getLong(FIELD_CREATED);
-            long expires = jsonObject.getLong(FIELD_EXPIRES);
-            String id = jsonObject.getString(FIELD_ID);
-            boolean liveMode = jsonObject.getBoolean(FIELD_LIVEMODE);
-            String object = jsonObject.getString(FIELD_OBJECT);
-            String secret = jsonObject.getString(FIELD_SECRET);
-
-            // Get the values from the associated objects array first element
-            JSONArray associatedObjectArray = jsonObject.getJSONArray(FIELD_ASSOCIATED_OBJECTS);
-            JSONObject typeObject = associatedObjectArray.getJSONObject(0);
-            String type = typeObject.getString(FIELD_TYPE);
-            String customerId = typeObject.getString(FIELD_ID);
-
-            return new EphemeralKey(
-                    created,
-                    customerId,
-                    expires,
-                    id,
-                    liveMode,
-                    object,
-                    secret,
-                    type);
-
-        } catch (JSONException ignored) {
-            return null;
+            return (TEphemeralKey)
+                    ephemeralKeyClass.getConstructor(JSONObject.class).newInstance(jsonObject);
```

Comment:
A little bit of reflection here will allow adding new type of Ephemeral keys in other modules